From 4db1edf73d3f16ceea5eaa13e1b8d1ba5d9cd922 Mon Sep 17 00:00:00 2001
Message-ID: <4db1edf73d3f16ceea5eaa13e1b8d1ba5d9cd922.1776946505.git.bertrand.marquis@arm.com>
In-Reply-To: <55d56a174a0f04520d4a1558630bb0b739074b13.1776946505.git.bertrand.marquis@arm.com>
References: <55d56a174a0f04520d4a1558630bb0b739074b13.1776946505.git.bertrand.marquis@arm.com>
From: Bertrand Marquis <bertrand.marquis@arm.com>
Date: Thu, 16 Apr 2026 19:30:45 +0200
Subject: [RFC PATCH 09/28] physmem: reject bounce buffers for direct-map-only
 address spaces

Make address_space_map() fail cleanly when a translation is not direct
and the AddressSpace is marked direct-map-only.

This prevents physmem from allocating bounce buffers in callers that
require a true direct mapping.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>
---
 system/physmem.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/system/physmem.c b/system/physmem.c
index b00684fc75cb..a849b78bed2b 100644
--- a/system/physmem.c
+++ b/system/physmem.c
@@ -3771,6 +3771,12 @@ void *address_space_map(AddressSpace *as,
     mr = flatview_translate(fv, addr, &xlat, &l, is_write, attrs);
 
     if (!memory_access_is_direct(mr, is_write, attrs)) {
+        if (as->direct_map_only) {
+            rcu_read_unlock();
+            *plen = 0;
+            return NULL;
+        }
+
         size_t used = qatomic_read(&as->bounce_buffer_size);
         for (;;) {
             hwaddr alloc = MIN(as->max_bounce_buffer_size - used, l);
-- 
2.53.0

